File: tests/macos/ClipboardBridge.m
#import <Metal/Metal.h>
#import <AppKit/AppKit.h>
#import <objc/runtime.h>
#include <assert.h>
#include <string.h>
#include "ClipboardBridge.h"

// Replace only the C bridge: the tests run the production AppKit handlers.
static uint8_t written[4096];
static size_t written_count;
static uint8_t tabs[8];
static size_t tab_count;
static MosttyScrollbar scrollbars[8];
static MosttyTab *write_tab;
static bool bracketed_paste;
static bool application_keypad;
static size_t keypad_queries;
static bool selection_active;
static uint32_t selection_end;
static bool word_selection;
static uint32_t word_col;
static id<MTLDevice> device;
static NSString *url_text;
static NSString *opened_url;
static bool hovered_url;
static bool url_open_success;
static IMP original_open_url;
static bool mouse_enabled;
static uint32_t mouse_count, mouse_action, mouse_button;
static int32_t mouse_x, mouse_y;

void clipboard_test_mouse_mode(bool enabled) { mouse_enabled = enabled; mouse_count = 0; }
uint32_t clipboard_test_mouse_count(void) { return mouse_count; }
uint32_t clipboard_test_mouse_action(void) { return mouse_action; }
uint32_t clipboard_test_mouse_button(void) { return mouse_button; }
int32_t clipboard_test_mouse_x(void) { return mouse_x; }
int32_t clipboard_test_mouse_y(void) { return mouse_y; }
bool mostty_tab_mouse_enabled(MosttyTab *tab) { return mouse_enabled; }
void mostty_tab_mouse(MosttyTab *tab, uint32_t action, uint32_t button, uint32_t mods, int32_t x, int32_t y) {
    write_tab = tab;
    mouse_count += 1; mouse_action = action; mouse_button = button; mouse_x = x; mouse_y = y;
}

static BOOL recordOpenURL(id workspace, SEL selector, NSURL *url) {
    opened_url = url.absoluteString;
    return url_open_success;
}

void clipboard_test_url(const char *url, bool open_success) {
    if (!original_open_url) {
        Method method = class_getInstanceMethod(NSWorkspace.class, @selector(openURL:));
        original_open_url = method_setImplementation(method, (IMP)recordOpenURL);
    }
    url_text = url ? [NSString stringWithUTF8String:url] : nil;
    opened_url = nil;
    url_open_success = open_success;
}
const char *clipboard_test_opened_url(void) { return opened_url.UTF8String; }
bool clipboard_test_hovered_url(void) { return hovered_url; }
void clipboard_test_restore_url_open(void) {
    if (original_open_url) {
        Method method = class_getInstanceMethod(NSWorkspace.class, @selector(openURL:));
        method_setImplementation(method, original_open_url);
        original_open_url = NULL;
    }
}

void clipboard_test_reset(bool bracketed) { written_count = 0; bracketed_paste = bracketed; }
void clipboard_test_keypad_mode(bool enabled) {
    application_keypad = enabled;
    keypad_queries = 0;
}
size_t clipboard_test_keypad_queries(void) { return keypad_queries; }
size_t clipboard_test_written(uint8_t *buf, size_t cap) {
    assert(written_count <= cap);
    memcpy(buf, written, written_count);
    return written_count;
}

MosttyTab *mostty_tab_create(uint32_t w, uint32_t h, float scale) {
    assert(tab_count < sizeof(tabs));
    scrollbars[tab_count] = (MosttyScrollbar){h / 20, 0, h / 20};
    return (MosttyTab *)&tabs[tab_count++];
}
MosttyTab *mostty_tab_create_with_launcher(uint32_t w, uint32_t h, float scale,
                                         const char *command, const char *directory) {
    return mostty_tab_create(w, h, scale);
}
MosttyTab *clipboard_test_created_tab(void) { return tab_count ? (MosttyTab *)&tabs[tab_count - 1] : NULL; }
MosttyTab *clipboard_test_write_tab(void) { return write_tab; }
void mostty_tab_destroy(MosttyTab *tab) {}
float mostty_config_background_opacity(void) { return 1; }
uint32_t mostty_config_render_interval_ms(void) { return 16; }
bool mostty_tab_apply_config(MosttyTab *tab) { return false; }
void *mostty_tab_metal_device(MosttyTab *tab) {
    if (!device) device = MTLCreateSystemDefaultDevice();
    return (__bridge void *)device;
}
intptr_t mostty_tab_read(MosttyTab *tab, uint8_t *buf, size_t cap) { return 0; }
void mostty_tab_feed(MosttyTab *tab, const uint8_t *ptr, size_t len) {}
void mostty_tab_write(MosttyTab *tab, const uint8_t *ptr, size_t len) {
    write_tab = tab;
    assert(len <= sizeof(written) - written_count);
    memcpy(written + written_count, ptr, len);
    written_count += len;
}
bool mostty_tab_set_surface(MosttyTab *tab, uint32_t w, uint32_t h, float scale, uint32_t *cols, uint32_t *rows) {
    *cols = w / 10; *rows = h / 20; return true;
}
void *mostty_tab_render(MosttyTab *tab, bool cursor, uint32_t *cols, uint32_t *rows) { return NULL; }
size_t mostty_tab_title(MosttyTab *tab, uint8_t *buf, size_t cap) { return 0; }
bool mostty_tab_poll_exit(MosttyTab *tab, int32_t *code) { return false; }
void mostty_tab_cell_size(MosttyTab *tab, uint32_t *w, uint32_t *h) { *w = 10; *h = 20; }
void mostty_tab_cursor(MosttyTab *tab, uint32_t *col, uint32_t *row) { *col = 0; *row = 0; }
bool mostty_tab_app_cursor_keys(MosttyTab *tab) { return false; }
bool mostty_tab_app_keypad(MosttyTab *tab) { keypad_queries += 1; return application_keypad; }
bool mostty_tab_bracketed_paste(MosttyTab *tab) { return bracketed_paste; }
MosttyScrollbar mostty_tab_scrollbar(MosttyTab *tab) {
    return tab ? scrollbars[(uint8_t *)tab - tabs] : (MosttyScrollbar){0};
}
void clipboard_test_scrollback(MosttyTab *tab, uint64_t total, uint64_t offset, uint64_t visible) {
    scrollbars[(uint8_t *)tab - tabs] = (MosttyScrollbar){total, offset, visible};
}
void mostty_tab_scroll_to_row(MosttyTab *tab, uint64_t row) {
    if (!tab) return;
    MosttyScrollbar *state = &scrollbars[(uint8_t *)tab - tabs];
    uint64_t max_offset = state->total > state->visible ? state->total - state->visible : 0;
    state->offset = MIN(row, max_offset);
}
void mostty_tab_scroll(MosttyTab *tab, int32_t rows) {
    MosttyScrollbar state = mostty_tab_scrollbar(tab);
    mostty_tab_scroll_to_row(tab, MAX(0, (int64_t)state.offset + rows));
}
void mostty_tab_scroll_to_bottom(MosttyTab *tab) { mostty_tab_scroll_to_row(tab, UINT64_MAX); }
void mostty_tab_set_selection(MosttyTab *tab, bool active, uint32_t sc, uint32_t sr, uint32_t ec, uint32_t er) {
    selection_active = active; selection_end = ec; word_selection = false;
}
bool mostty_tab_select_word(MosttyTab *tab, uint32_t col, uint32_t row) {
    selection_active = true; word_selection = true; word_col = col;
    return true;
}
bool mostty_tab_hover_url(MosttyTab *tab, bool active, uint32_t col, uint32_t row) {
    hovered_url = active && url_text != nil && col >= 10 && col < 30 && row == 0;
    return hovered_url;
}
size_t mostty_tab_url_at(MosttyTab *tab, uint32_t col, uint32_t row, uint8_t *buf, size_t cap) {
    if (!url_text || col < 10 || col >= 30 || row != 0) return 0;
    const char *text = url_text.UTF8String;
    size_t len = strlen(text);
    if (len > cap) return 0;
    memcpy(buf, text, len);
    return len;
}
size_t mostty_tab_selection_text(MosttyTab *tab, uint8_t *buf, size_t cap) {
    assert(selection_active);
    const char *text = word_selection ? (word_col == 8 ? "x" : "hello") : (selection_end == 4 ? "hello" : "hel");
    size_t len = strlen(text);
    if (cap == 0) return len;
    assert(len <= cap);
    memcpy(buf, text, len);
    return len;
}

@interface ClipboardDragInfo : NSObject <NSDraggingInfo>
@property(strong) NSPasteboard *draggingPasteboard;
@property NSDragOperation draggingSourceOperationMask;
@property NSDraggingFormation draggingFormation;
@property BOOL animatesToDestination;
@property NSInteger numberOfValidItemsForDrop;
@end

@implementation ClipboardDragInfo
- (NSWindow *)draggingDestinationWindow { return nil; }
- (NSPoint)draggingLocation { return NSZeroPoint; }
- (NSPoint)draggedImageLocation { return NSZeroPoint; }
- (NSImage *)draggedImage { return nil; }
- (id)draggingSource { return nil; }
- (NSInteger)draggingSequenceNumber { return 0; }
- (NSSpringLoadingHighlight)springLoadingHighlight { return NSSpringLoadingHighlightNone; }
- (void)resetSpringLoading {}
- (void)slideDraggedImageTo:(NSPoint)point {}
- (NSArray<NSString *> *)namesOfPromisedFilesDroppedAtDestination:(NSURL *)destination { return nil; }
- (void)enumerateDraggingItemsWithOptions:(NSDraggingItemEnumerationOptions)options
                                forView:(NSView *)view classes:(NSArray<Class> *)classes
                          searchOptions:(NSDictionary<NSPasteboardReadingOptionKey, id> *)searchOptions
                             usingBlock:(void (^)(NSDraggingItem *, NSInteger, BOOL *))block {}
@end

id<NSDraggingInfo> clipboard_test_drag(NSPasteboard *pasteboard, NSDragOperation operation) {
    ClipboardDragInfo *info = [ClipboardDragInfo new];
    info.draggingPasteboard = pasteboard;
    info.draggingSourceOperationMask = operation;
    return info;
}

File: tests/macos/ClipboardBridge.swift
import AppKit
import Metal
import ObjectiveC

// Replace only the C bridge: the tests run the production AppKit handlers.
private var written: [UInt8] = []
private let tabs = UnsafeMutablePointer<UInt8>.allocate(capacity: 8)
private var tabCount = 0
private var scrollbars = [MosttyScrollbar](repeating: MosttyScrollbar(), count: 8)
private var writeTab: OpaquePointer?
private var bracketedPaste = false
private var applicationKeypad = false
private var keypadQueries = 0
private var selectionActive = false
private var selectionEnd: UInt32 = 0
private var wordSelection = false
private var wordCol: UInt32 = 0
private var device: MTLDevice?
private var urlText: String?
private var openedURL: String?
private var hoveredURL = false
private var urlOpenSuccess = false
private var originalOpenURL: IMP?
private var mouseEnabled = false
private var mouseCount: UInt32 = 0
private var mouseAction: UInt32 = 0
private var mouseButton: UInt32 = 0
private var mouseX: Int32 = 0
private var mouseY: Int32 = 0

func clipboard_test_mouse_mode(_ enabled: Bool) { mouseEnabled = enabled; mouseCount = 0 }
func clipboard_test_mouse_count() -> UInt32 { mouseCount }
func clipboard_test_mouse_action() -> UInt32 { mouseAction }
func clipboard_test_mouse_button() -> UInt32 { mouseButton }
func clipboard_test_mouse_x() -> Int32 { mouseX }
func clipboard_test_mouse_y() -> Int32 { mouseY }

@_cdecl("mostty_tab_mouse_enabled")
func stubMouseEnabled(_ tab: OpaquePointer?) -> Bool { mouseEnabled }
@_cdecl("mostty_tab_mouse")
func stubMouse(_ tab: OpaquePointer?, _ action: UInt32, _ button: UInt32, _ mods: UInt32, _ x: Int32, _ y: Int32) {
    writeTab = tab
    mouseCount += 1; mouseAction = action; mouseButton = button; mouseX = x; mouseY = y
}

private let recordOpenURL: @convention(c) (AnyObject, Selector, NSURL) -> Bool = { _, _, url in
    openedURL = url.absoluteString
    return urlOpenSuccess
}
func clipboard_test_url(_ url: String?, _ openSuccess: Bool) {
    if originalOpenURL == nil {
        let method = class_getInstanceMethod(NSWorkspace.self, #selector(NSWorkspace.open(_:)))!
        originalOpenURL = method_setImplementation(method, unsafeBitCast(recordOpenURL, to: IMP.self))
    }
    urlText = url
    openedURL = nil
    urlOpenSuccess = openSuccess
}
func clipboard_test_opened_url() -> String? { openedURL }
func clipboard_test_hovered_url() -> Bool { hoveredURL }
func clipboard_test_restore_url_open() {
    if let original = originalOpenURL {
        let method = class_getInstanceMethod(NSWorkspace.self, #selector(NSWorkspace.open(_:)))!
        method_setImplementation(method, original)
        originalOpenURL = nil
    }
}
func clipboard_test_reset(_ bracketed: Bool) { written.removeAll(); bracketedPaste = bracketed }
func clipboard_test_keypad_mode(_ enabled: Bool) { applicationKeypad = enabled; keypadQueries = 0 }
func clipboard_test_keypad_queries() -> Int { keypadQueries }
func clipboard_test_written(_ buffer: UnsafeMutablePointer<UInt8>, _ capacity: Int) -> Int {
    precondition(written.count <= capacity)
    buffer.update(from: written, count: written.count)
    return written.count
}

@_cdecl("mostty_tab_create")
func stubCreate(_ width: UInt32, _ height: UInt32, _ scale: Float) -> OpaquePointer? {
    precondition(tabCount < 8)
    scrollbars[tabCount] = MosttyScrollbar(total: UInt64(height / 20), offset: 0, visible: UInt64(height / 20))
    defer { tabCount += 1 }
    return OpaquePointer(tabs.advanced(by: tabCount))
}
@_cdecl("mostty_tab_create_with_launcher")
func stubCreateWithLauncher(_ width: UInt32, _ height: UInt32, _ scale: Float,
                           _ command: UnsafePointer<CChar>?, _ directory: UnsafePointer<CChar>?) -> OpaquePointer? {
    stubCreate(width, height, scale)
}
func clipboard_test_created_tab() -> OpaquePointer? { tabCount > 0 ? OpaquePointer(tabs.advanced(by: tabCount - 1)) : nil }
func clipboard_test_write_tab() -> OpaquePointer? { writeTab }
@_cdecl("mostty_tab_destroy")
func stubDestroy(_ tab: OpaquePointer?) {}
@_cdecl("mostty_config_background_opacity")
func stubBackgroundOpacity() -> Float { 1 }
@_cdecl("mostty_config_copy_tabbar_font")
func stubCopyTabbarFont() -> UnsafeMutableRawPointer? {
    Unmanaged.passRetained(NSFont(name: "Menlo", size: 13)!).toOpaque()
}
@_cdecl("mostty_config_render_interval_ms")
func stubRenderInterval() -> UInt32 { 16 }
@_cdecl("mostty_tab_apply_config")
func stubApplyConfig(_ tab: OpaquePointer?) -> Bool { false }
@_cdecl("mostty_tab_metal_device")
func stubMetalDevice(_ tab: OpaquePointer?) -> UnsafeMutableRawPointer? {
    if device == nil { device = MTLCreateSystemDefaultDevice() }
    return device.map { Unmanaged.passUnretained($0 as AnyObject).toOpaque() }
}
@_cdecl("mostty_tab_read")
func stubRead(_ tab: OpaquePointer?, _ buffer: UnsafeMutablePointer<UInt8>?, _ capacity: Int) -> Int {
    // A live idle session times out; EOF would asynchronously close its pane.
    usleep(10000)
    return -2
}
@_cdecl("mostty_tab_feed")
func stubFeed(_ tab: OpaquePointer?, _ pointer: UnsafePointer<UInt8>?, _ length: Int) {}
@_cdecl("mostty_tab_write")
func stubWrite(_ tab: OpaquePointer?, _ pointer: UnsafePointer<UInt8>?, _ length: Int) {
    writeTab = tab
    precondition(length <= 4096 - written.count)
    written.append(contentsOf: UnsafeBufferPointer(start: pointer, count: length))
}
@_cdecl("mostty_tab_set_surface")
func stubSetSurface(_ tab: OpaquePointer?, _ width: UInt32, _ height: UInt32, _ scale: Float,
                    _ cols: UnsafeMutablePointer<UInt32>?, _ rows: UnsafeMutablePointer<UInt32>?) -> Bool {
    cols?.pointee = width / 10; rows?.pointee = height / 20
    return true
}
@_cdecl("mostty_tab_render")
func stubRender(_ tab: OpaquePointer?, _ cursor: Bool, _ textBlinkOn: Bool,
                _ cols: UnsafeMutablePointer<UInt32>?, _ rows: UnsafeMutablePointer<UInt32>?) -> UnsafeMutableRawPointer? { nil }
@_cdecl("mostty_tab_title")
func stubTitle(_ tab: OpaquePointer?, _ buffer: UnsafeMutablePointer<UInt8>?, _ capacity: Int) -> Int { 0 }
@_cdecl("mostty_tab_poll_exit")
func stubPollExit(_ tab: OpaquePointer?, _ code: UnsafeMutablePointer<Int32>?) -> Bool { false }
@_cdecl("mostty_tab_cell_size")
func stubCellSize(_ tab: OpaquePointer?, _ width: UnsafeMutablePointer<UInt32>?, _ height: UnsafeMutablePointer<UInt32>?) {
    width?.pointee = 10; height?.pointee = 20
}
@_cdecl("mostty_tab_cursor")
func stubCursor(_ tab: OpaquePointer?, _ col: UnsafeMutablePointer<UInt32>?, _ row: UnsafeMutablePointer<UInt32>?) {
    col?.pointee = 0; row?.pointee = 0
}
@_cdecl("mostty_tab_app_cursor_keys")
func stubAppCursorKeys(_ tab: OpaquePointer?) -> Bool { false }
@_cdecl("mostty_tab_app_keypad")
func stubAppKeypad(_ tab: OpaquePointer?) -> Bool { keypadQueries += 1; return applicationKeypad }
@_cdecl("mostty_tab_bracketed_paste")
func stubBracketedPaste(_ tab: OpaquePointer?) -> Bool { bracketedPaste }
@_cdecl("mostty_tab_scrollbar")
func stubScrollbar(_ tab: OpaquePointer?) -> MosttyScrollbar {
    guard let tab else { return MosttyScrollbar() }
    return scrollbars[tabs.distance(to: UnsafeMutablePointer<UInt8>(tab))]
}
func clipboard_test_scrollback(_ tab: OpaquePointer, _ total: UInt64, _ offset: UInt64, _ visible: UInt64) {
    scrollbars[tabs.distance(to: UnsafeMutablePointer<UInt8>(tab))] = MosttyScrollbar(total: total, offset: offset, visible: visible)
}
@_cdecl("mostty_tab_scroll_to_row")
func stubScrollToRow(_ tab: OpaquePointer?, _ row: UInt64) {
    guard let tab else { return }
    let index = tabs.distance(to: UnsafeMutablePointer<UInt8>(tab))
    let state = scrollbars[index]
    scrollbars[index].offset = min(row, state.total > state.visible ? state.total - state.visible : 0)
}
@_cdecl("mostty_tab_scroll")
func stubScroll(_ tab: OpaquePointer?, _ rows: Int32) {
    let state = stubScrollbar(tab)
    stubScrollToRow(tab, UInt64(max(0, Int64(state.offset) + Int64(rows))))
}
@_cdecl("mostty_tab_scroll_to_bottom")
func stubScrollToBottom(_ tab: OpaquePointer?) { stubScrollToRow(tab, UInt64.max) }
@_cdecl("mostty_tab_set_selection")
func stubSetSelection(_ tab: OpaquePointer?, _ active: Bool, _ sc: UInt32, _ sr: UInt32, _ ec: UInt32, _ er: UInt32) {
    selectionActive = active; selectionEnd = ec; wordSelection = false
}
@_cdecl("mostty_tab_select_word")
func stubSelectWord(_ tab: OpaquePointer?, _ col: UInt32, _ row: UInt32) -> Bool {
    selectionActive = true; wordSelection = true; wordCol = col
    return true
}
@_cdecl("mostty_tab_hover_url")
func stubHoverURL(_ tab: OpaquePointer?, _ active: Bool, _ col: UInt32, _ row: UInt32) -> Bool {
    hoveredURL = active && urlText != nil && col >= 10 && col < 30 && row == 0
    return hoveredURL
}
@_cdecl("mostty_tab_url_at")
func stubURLAt(_ tab: OpaquePointer?, _ col: UInt32, _ row: UInt32,
               _ buffer: UnsafeMutablePointer<UInt8>?, _ capacity: Int) -> Int {
    guard let text = urlText, col >= 10, col < 30, row == 0 else { return 0 }
    let bytes = Array(text.utf8)
    guard bytes.count <= capacity else { return 0 }
    buffer?.update(from: bytes, count: bytes.count)
    return bytes.count
}
@_cdecl("mostty_tab_selection_text")
func stubSelectionText(_ tab: OpaquePointer?, _ buffer: UnsafeMutablePointer<UInt8>?, _ capacity: Int) -> Int {
    precondition(selectionActive)
    let text = wordSelection ? (wordCol == 8 ? "x" : "hello") : (selectionEnd == 4 ? "hello" : "hel")
    let bytes = Array(text.utf8)
    if capacity == 0 { return bytes.count }
    precondition(bytes.count <= capacity)
    buffer?.update(from: bytes, count: bytes.count)
    return bytes.count
}

private final class ClipboardDragInfo: NSObject, NSDraggingInfo {
    let draggingPasteboard: NSPasteboard
    let draggingSourceOperationMask: NSDragOperation
    var draggingFormation: NSDraggingFormation = .default
    var animatesToDestination = false
    var numberOfValidItemsForDrop = 0
    var draggingDestinationWindow: NSWindow? { nil }
    var draggingLocation: NSPoint { .zero }
    var draggedImageLocation: NSPoint { .zero }
    var draggedImage: NSImage? { nil }
    var draggingSource: Any? { nil }
    var draggingSequenceNumber: Int { 0 }
    var springLoadingHighlight: NSSpringLoadingHighlight { .none }

    init(_ pasteboard: NSPasteboard, _ operation: NSDragOperation) {
        draggingPasteboard = pasteboard
        draggingSourceOperationMask = operation
        super.init()
    }
    func resetSpringLoading() {}
    func slideDraggedImage(to screenPoint: NSPoint) {}
    override func namesOfPromisedFilesDropped(atDestination dropDestination: URL) -> [String]? { nil }
    func enumerateDraggingItems(options enumOpts: NSDraggingItemEnumerationOptions = [], for view: NSView?,
                                classes classArray: [AnyClass], searchOptions: [NSPasteboard.ReadingOptionKey: Any] = [:],
                                using block: (NSDraggingItem, Int, UnsafeMutablePointer<ObjCBool>) -> Void) {}
}
func clipboard_test_drag(_ pasteboard: NSPasteboard, _ operation: NSDragOperation) -> NSDraggingInfo {
    ClipboardDragInfo(pasteboard, operation)
}
